Adding keyboard shortcuts to navigate on the Block Insert Menu

youknowriad opened the pull request:
refs #87 

This PRs adds keyboards shortcuts to navigate through the insert block menu.

 - `/` to open the menu (working only when the editor is not focused for now)
 - `Enter` to insert the current active block (Obviously, it does nothing aside closing the menu for now)
 - Navigate with arrows to select blocks
 - Type text even if the input is not focused

Obviously, this PR adds so complexity to the code. Should we consider adding some organisation, modules?

**Notes:**

I tried to allow opening the menu when we type `/` on an empty block, but it's not easy to get the active block with the current implementation. Maybe in another PR.

youknowriad committed:
Adding keyboard shortcuts to navigate on the Block Insert Menu

timmyc commented on blocks.js:
```diff
@@ -45,44 +45,52 @@ var config = {
 			label: 'Paragraph',
 			icon: '<svg height="24" width="24" class="type-icon-paragraph" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path id="path-1_2_" class="st0" d="M13 5h2v16h2V5h2V3h-6.7.2-3C6.5 3 4 5.5 4 8.5S6.5 14 9.5 14H11v7h2v-7h-.5.5V5z"/><path class="st1" d="M9.5 3C6.5 3 4 5.5 4 8.5S6.5 14 9.5 14H11v7h2V5h2v16h2V5h2V3H9.5z"/></svg>',
-			categories: [ 'frequent' ]
+			category: 'frequent'
```
Couldn't some blocks appear in multiple categories?  I'm guessing that was the initial intention of having an array here.

youknowriad replied:
Yes, actually I did this in a previous Pr. but as @jasmussen said is his comment, we should stick with one category for now https://github.com/WordPress/gutenberg/issues/87#issuecomment-280649417
which simplifies a lot the navigation using arrows

jasmussen commented:
Solid work. 

Ran into a few issues:

![feb-17-2017 16-57-04](https://cloud.githubusercontent.com/assets/1204802/23072488/1de3d686-f532-11e6-866c-b743041e2764.gif)

Couldn't get / to invoke the menu, though I suspect it's because I'm on a european keyboard. 

The insert menu scrolled a bit while using the search feature, that felt a little jumpy somehow. Feels like the scrollbar should be glued to the top as you start typing — think of it as the entire categorized & curated window being replaced just by search results. Try it in the OSX emoji keyboard accessible by ^⌘ Space, then search for "star". 

If I click the plus button first, then start typing, works great. If I put the text cursor inside text first, then search, then I'm actually typing in the text (which closes the search field). 

One iffy thing which I suspect will cause us some headaches — if you arrow key to the bottom of the inserter, and keep going down, you'll start scrolling the page instead. 

Finally, and I'll pose this more as a question for the group than as a critique: what should happen when you arrow-key to the bottom of the left column of blocks and keep going? Should it stop, or should it start over from the top right column? Like this key direciton:

![screen shot 2017-02-17 at 17 02 25](https://cloud.githubusercontent.com/assets/1204802/23072689/f1e2c47e-f532-11e6-9cc2-32c32535ef9b.png)

I realize this is still mostly a prototype, but feels like this particular interaction is important to get right.

iseulde commented:
You might want to have a look at our text pattern plugin to detect normal characters: https://github.com/WordPress/wordpress-develop/blob/master/src/wp-includes/js/tinymce/plugins/wptextpattern/plugin.js
You'll probably have to look into the text node on `keyup`.

youknowriad committed:
Changes per review

youknowriad commented:
I updated the PR, I changed the way I track the `/` key press ( let me know if it works better for you now @jasmussen ). I also fixed some scrolling issues. What do you think about the behaviour now?

jasmussen commented:
Nice! Feels much more solid. 

However now I can't press enter or type in any of the text boxes. 

Just to clarify — the thing we're mimicking here is slack. Try opening Slackbot and start typing `/aw`(and it'll soon suggest `/away`.

youknowriad commented:
@jasmussen Oh! I see, I understand better now :) They should be a bit harder to achieve in this "non-scalable" prototype but I'll try to figure it out

jasmussen commented:
So long as you can actually type in the textblocks, there's probably no need to take it _all the way_. The idea is to convey a feeling, at least until we get into the technical prototypes.

youknowriad committed:
Changes per review

youknowriad commented:
@jasmussen would you mind taking a look at this again?

I updated and now you can trigger the menu when you type `/` on an empty line on the editor and after selecting the block, you can continue typing inside the text block.

You still have the possibility to trigger the menu without even focusing the editor.

jasmussen commented:
SOLID work. Love it!

Definitely good to merge for the UI prototype. 

Future improvements (some dependant on other tickets) could include:

- show insertion menu in context of the newline on which the `/` was typed
- actually show the search term on the newline (the newline is your command line) until you press enter. E.g. type `/ima`, press enter, and an image is inserted

See also https://github.com/WordPress/gutenberg/blob/master/mockups/Insert%20Between,%20Keyboard.png

youknowriad commented:
@mtias would you mind a technical Review for this or maybe we don't care that much since it's a prototype?

mtias commented:
The block-switcher between paragraph, quote, and heading seems to be a bit broken. Not sure if it's specific to this branch or also present in master.

If it's not a regression here, feel free to merge.

youknowriad commented:
In my testing, sometimes the block-switcher does not switch as expected, we have to try several times and I was able to reproduce in master as well. Merging!
